notebooks/analyze_looping_across_lending_protocols.ipynb

```diff
@@ -424,45 +397,34 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "# Helper function to parse token amounts\n",
-    "def parse_token_amounts(token_str):\n",
-    "    tokens = []\n",
-    "    if token_str:\n",
-    "        token_pairs = token_str.split(', ')\n",
-    "        for token_pair in token_pairs:\n",
-    "            parts = token_pair.split(': ')\n",
-    "            if len(parts) == 2:\n",
-    "                token_name, amount = parts\n",
-    "                tokens.append((token_name, float(amount)))\n",
-    "    return tokens\n",
-    "\n",
-    "\n",
-    "# Function to extract user actions from the dataframes\n",
-    "def extract_user_actions(dataframes, tokens):\n",
-    "    user_actions = defaultdict(list)\n",
-    "    for protocol, df in dataframes.items():\n",
-    "        for _, row in df.iterrows():\n",
-    "            user = row['User']\n",
-    "            collateral_tokens = parse_token_amounts(row['Collateral'])\n",
-    "            debt_tokens = parse_token_amounts(row['Debt'])\n",
-    "\n",
-    "            # Store collateral actions\n",
-    "            for col_token, col_amount in collateral_tokens:\n",
-    "                if col_token in tokens:\n",
-    "                    user_actions[user].append(('collateral', col_token, col_amount, protocol))\n",
-    "\n",
-    "            # Store debt actions\n",
-    "            for debt_token, debt_amount in debt_tokens:\n",
-    "                if debt_token in tokens:\n",
-    "                    user_actions[user].append(('debt', debt_token, debt_amount, protocol))\n",
-    "    return user_actions\n",
-    "\n",
-    "# Extract user actions\n",
-    "user_actions = extract_user_actions(dataframes, tokens)\n",
-    "\n",
+    "# Plot the results in a pie chart\n",
+    "labels = [F\"Total Users: {total_users}\", f\"Looping Users: {looping_users}\"]\n",
+    "sizes = [total_users, looping_users]\n",
+    "colors = ['#66b3ff', '#ff9999']\n",
+    "explode = (0.1, 0)  # explode the 2nd slice (Looping users)\n",
     "\n",
+    "plt.figure(figsize=(8, 8))\n",
+    "plt.pie(sizes, explode=explode, labels=labels, colors=colors, autopct='%1.1f%%', shadow=True, startangle=140)\n",
+    "plt.title('Comparison of Total Users vs Users that has completed the loop')\n",
+    "plt.axis('equal')  # Equal aspect ratio ensures that pie is drawn as a circle.\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## What is the total dollar amount of tokens used in these loops? How much are the deposits multiplicated?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
     "# Helper function to check if a loop exists and calculate total dollar amounts\n",
-    "def has_completed_loop(actions):\n",
+    "def has_completed_loop(actions: List[Tuple[str, str, float, str]]) -> float:\n",
```

Comment:
This function checks (in a complicated way), whether the collateral and debt actions contain the same token. But I don't see how it checks for the "loop" occurrence. Also, based on how the "loop" is defined, you need to use the information about the protocol as well.

Reply:
Done! ✅